### PR TITLE
Use the file resolver in LazyPublishArtifact

### LINKED ISSUE #15711: Unable to use a Provider<java.nio.file.Path> as an artifact for the maven-publish plugin
Using a `Provider<java.nio.file.Path>` as an artifact for the `maven-publish`-Plugin does not work in Gradle `6.6` anymore. Using a Path (`sun.nio.fs.WindowsPath`) instance directly still works.

### Expected Behavior

The file `my-artifact.txt` should be published into the local maven repository.

### Current Behavior

You get the following error:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generatePomFileForArtifactoryPublication'.
> Cannot convert provided value (my-artifact.txt) to a file.
```

### Context

This works in Gradle `6.5.1` and below, but stops working in Gradle `6.6`. I also tested all newer versions up to `6.7.1`, they are all affected.

May be related to #11054.

### Steps to Reproduce 

Execute `gradlew publishToMavenLocal` in this project: [gradle-issue-15711.zip](https://github.com/gradle/gradle/files/5781395/gradle-issue-15711.zip)

For reference here is the `build.gradle`:

```gradle
apply plugin: 'maven-publish'

wrapper { gradleVersion = '6.6' }

group = 'org.gradle'
ext.artifactId = 'issue-15711'
version = '1.0.0'

publishing {
  publications {
    artifactory(MavenPublication) {
      artifactId = project.artifactId
      // Works in Gradle 6.5.1 and Gradle 6.6
//      artifact java.nio.file.Path.of('my-artifact.txt')

      // Works in Gradle 6.5.1 but not Gradle 6.6
      artifact provider { java.nio.file.Path.of('my-artifact.txt') }
    }
  }
}
```

### Your Environment

Build scan URL: https://gradle.com/s/qxutkwm5vnfdq

```
user@PC MINGW64 /c/.../gradle-issue-15711
$ ./gradlew -v

------------------------------------------------------------
Gradle 6.6
------------------------------------------------------------

Build time:   2020-08-10 22:06:19 UTC
Revision:     d119144684a0c301aea027b79857815659e431b9

Kotlin:       1.3.72
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.8 compiled on May 10 2020
JVM:          1.8.0_162 (Oracle Corporation 25.162-b12)
OS:           Windows 7 6.1 amd64
```

### PR DESCRIPTION
**Pull request by octylFractal**
This gives us many of these conversions for free, reducing duplicated code. This also fixes #15711.

The bulk of the change is `LazyPublishArtifact.java`, everything else is dependency injection or test fixing.

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle help

**Comment by big-guy** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java [hunk 1]
```diff
@@ -180,6 +187,39 @@ public static void configureDocumentationVariantWithArtifact(
         ConfigurationContainer configurations,
         TaskContainer tasks,
         ObjectFactory objectFactory
+    ) {
+        DeprecationLogger.deprecateInternalApi("configureDocumentationVariantWithArtifact (no FileResolver)")
+            .replaceWith("configureDocumentationVariantWithArtifact (with FileResolver)")
+            .willBeRemovedInGradle8()
+            .undocumented()
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java:194]: same here

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by octylFractal** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge

**Comment by octylFractal** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge

**Review by big-guy — APPROVED** [review]

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle cancel

**Comment by octylFractal** [comment]
@bot-gradle help

**Comment by octylFractal** [comment]
@bot-gradle test

**Comment by octylFractal** [comment]
@bot-gradle test this

**Comment by octylFractal** [comment]
@bot-gradle test this

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Pull request merged**